ocaml/xapi/xapi_sr.ml

```diff
@@ -183,11 +183,43 @@ let probe ~__context ~host ~device_config ~_type ~sm_config =
 
   transform_storage_exn
     (fun () ->
-       match Client.SR.probe ~dbg ~queue ~device_config ~sm_config with
-       | Raw x -> x
-       | Probe _ as x -> Xmlrpc.to_string (rpc_of_probe_result x)
+       Client.SR.probe ~dbg ~queue ~device_config ~sm_config |> f
     )
 
+let probe = call_probe ~f:(
+    function
+    | Storage_interface.Raw x -> x
+    | Storage_interface.Probe probe_results ->
+      (* Here we try to mimic the XML document structure returned in the Raw
+         case by SMAPIv1, for backwards-compatibility *)
```

Comment:
Wouldn't it be better to add an optional version to the probe call, and when called with version say `>= 2` do not perform any conversion, so that new code can actually benefit from a structured safe response?

Reply:
Hmm, not sure, Jon said we should have a new `SR.probe_ext` call, that is the one that returns the structured data, in a way that is reflected by ocaml's type system too.

Reply:
We propose to introduce  a new API, probe_ext with the structed reply: https://github.com/xapi-project/xen-api/blob/feature/REQ477/probe/ocaml/idl/datamodel.ml#L1869

Reply:
I'm fine with that,I am happy to see there will be a new thing that uses a reasonable response format and that will be used by new code

Reply:
It's always very hard to change the return type of an API call, sadly, and we know there are people out there who do actually use the probe call, so we don't want to break it unnecessarily.